### DISCUSSION TITLE
What zospy type for ZOSAPI.Tools.RayTrace.OPDMode?

### DISCUSSION THREAD

**Discussion by mitkoge** [MREYE-LUMC/ZOSPy#53] [Q&A] [answered]
Hi, 
my appreciations for the great work done.

I try to convert the standard example PythonStandalone_22_seq_spot_diagram.py for batch ray tracing.

I have trouble with the following statement:

`normUnPolData.AddRay(waveNumber, hx, hy_ary[field - 1], px, py, Enum.Parse(ZOSAPI.Tools.RayTrace.OPDMode, "None"))
`
i translated it as 
`normUnPolData.AddRay(waveNumber, hx, hy_ary[field - 1], px, py, zp.constants.Tools.RayTrace.OPDMode[0])
`
but have no idea how to handle this Enum stuff.
I guess it has to translate the "None" to the right OPDmode function. 
But what type i have to use instead of  ZOSAPI.Tools.RayTrace.OPDMode?

best regards
Dimiter

**Comment by crnh** [comment, marked as answer]
Hi! First, thank you for using ZOSPy!

There are two ways to access ZOS-API constants from ZOSPy, and your method is one of them. `zp.constants.Tools.RayTrace.OPDMode[0]` will give you `ZOSAPI.Tools.RayTrace.OPDMode.None`.

Alternatively, and this is the preferred way, you can use `zp.constants.Tools.RayTrace.OPDMode.None_` (note the underscore at the end). ZOSPy's constants are designed to reflect the behavior of the ZOS-API Enums as much as possible, so you can just access their values as attributes, e.g. `ZOSAPI.Tools.RayTrace.OPDMode.Current`. However, this is not possible with `None`, because that name has a different meaning in Python. We therefore suffixed all occurences of `None` with an underscore, i.e. `None_`. The OpticStudio documentation uses a different syntax only for these special cases (`Enum.Parse(ZOSAPI.Tools.RayTrace.OPDMode, "None")`). You probably already discovered that this syntax does not work in ZOSPy, because ZOSPy's constants aren't Enums under the hood, although they are linked to the ZOS-API Enums.
